arch/Mips/MipsMapping.c

```diff
@@ -1,1071 +1,454 @@
+static void Mips_set_detail_op_imm(MCInst *MI, unsigned OpNum, int64_t Imm)
+{
+	if (!detail_is_set(MI))
+		return;
 
-	// architecture-specific groups
-	{ MIPS_GRP_BITCOUNT, "bitcount" },
-	{ MIPS_GRP_DSP, "dsp" },
-	{ MIPS_GRP_DSPR2, "dspr2" },
-	{ MIPS_GRP_FPIDX, "fpidx" },
-	{ MIPS_GRP_MSA, "msa" },
-	{ MIPS_GRP_MIPS32R2, "mips32r2" },
-	{ MIPS_GRP_MIPS64, "mips64" },
-	{ MIPS_GRP_MIPS64R2, "mips64r2" },
-	{ MIPS_GRP_SEINREG, "seinreg" },
-	{ MIPS_GRP_STDENC, "stdenc" },
-	{ MIPS_GRP_SWAP, "swap" },
-	{ MIPS_GRP_MICROMIPS, "micromips" },
-	{ MIPS_GRP_MIPS16MODE, "mips16mode" },
-	{ MIPS_GRP_FP64BIT, "fp64bit" },
-	{ MIPS_GRP_NONANSFPMATH, "nonansfpmath" },
-	{ MIPS_GRP_NOTFP64BIT, "notfp64bit" },
-	{ MIPS_GRP_NOTINMICROMIPS, "notinmicromips" },
-	{ MIPS_GRP_NOTNACL, "notnacl" },
+	if (doing_mem(MI)) {
+		Mips_set_detail_op_mem_disp(MI, OpNum, Imm);
+		return;
+	}
 
-	{ MIPS_GRP_NOTMIPS32R6, "notmips32r6" },
-	{ MIPS_GRP_NOTMIPS64R6, "notmips64r6" },
-	{ MIPS_GRP_CNMIPS, "cnmips" },
+	Mips_get_detail_op(MI, 0)->type = MIPS_OP_IMM;
+	Mips_get_detail_op(MI, 0)->imm = Imm;
+	Mips_get_detail_op(MI, 0)->access = map_get_op_access(MI, OpNum);
+	Mips_inc_op_count(MI);
+}
 
-	{ MIPS_GRP_MIPS32, "mips32" },
-	{ MIPS_GRP_MIPS32R6, "mips32r6" },
-	{ MIPS_GRP_MIPS64R6, "mips64r6" },
+static void Mips_set_detail_op_reg(MCInst *MI, unsigned OpNum, mips_reg Reg)
+{
+	if (!detail_is_set(MI))
+		return;
 
-	{ MIPS_GRP_MIPS2, "mips2" },
-	{ MIPS_GRP_MIPS3, "mips3" },
-	{ MIPS_GRP_MIPS3_32, "mips3_32"},
-	{ MIPS_GRP_MIPS3_32R2, "mips3_32r2" },
+	if (doing_mem(MI)) {
+		Mips_set_detail_op_mem_reg(MI, OpNum, Reg);
+		return;
+	}
 
-	{ MIPS_GRP_MIPS4_32, "mips4_32" },
-	{ MIPS_GRP_MIPS4_32R2, "mips4_32r2" },
-	{ MIPS_GRP_MIPS5_32R2, "mips5_32r2" },
+	assert((map_get_op_type(MI, OpNum) & ~CS_OP_MEM) == CS_OP_REG);
 
-	{ MIPS_GRP_GP32BIT, "gp32bit" },
-	{ MIPS_GRP_GP64BIT, "gp64bit" },
-};
-#endif
+	Mips_get_detail_op(MI, 0)->type = MIPS_OP_REG;
+	Mips_get_detail_op(MI, 0)->reg = Reg;
+	Mips_get_detail_op(MI, 0)->access = map_get_op_access(MI, OpNum);
+	Mips_inc_op_count(MI);
+}
 
-const char *Mips_group_name(csh handle, unsigned int id)
+static void Mips_set_detail_op_operand(MCInst *MI, unsigned OpNum)
 {
-#ifndef CAPSTONE_DIET
-	return id2name(group_name_maps, ARR_SIZE(group_name_maps), id);
-#else
-	return NULL;
-#endif
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		Mips_set_detail_op_imm(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else if (op_type == CS_OP_REG) {
+		Mips_set_detail_op_reg(MI, OpNum,
+					    MCInst_getOpVal(MI, OpNum));
+	} else
+		printf("Operand type %d not handled!\n", op_type);
+}
+
+static void Mips_set_detail_op_branch(MCInst *MI, unsigned OpNum)
+{
+	cs_op_type op_type = map_get_op_type(MI, OpNum) & ~CS_OP_MEM;
+	if (op_type == CS_OP_IMM) {
+		uint64_t Target = (uint64_t)MCInst_getOpVal(MI, OpNum);
+		Mips_set_detail_op_imm(MI, OpNum, Target + MI->address);
```

Comment:
You can check for the `csh->PrintBranchImmNotAsAddress` option here and only optionally add the address.

Also please rename the `CS_OPT_NO_BRANCH_OFFSET` option (it sets this flag). The naming was horrible chosen by me.

Reply:
Ouh, but the immediate offset is a `int64_t`. Not an `uint64_t`!. The LLVM code does:
```cpp
    uint64_t Target = Address + Op.getImm();
```

But `getImm(Op)` returns an `int64_t` (also our C version).

Reply:
this is a copy of what the printing instructions do. otherwise i need to use args to get this value (`address`).

Reply:
Sorry, I got confused about it while thinking about. It doesn't matter of course, both are 64bit